dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java

```diff
@@ -340,101 +437,76 @@ public ImportSummaries addEvents( List<Event> events, ImportOptions importOption
-                return new ImportSummary( ImportStatus.ERROR, "Tracked entity instance: " + entityInstance.getUid()
-                    + " is not enrolled in program: " + program.getUid() ).setReference( event.getEvent() )
-                    .incrementIgnored();
-            }
-            else if ( programInstances.size() > 1 )
-            {
-                return new ImportSummary( ImportStatus.ERROR, "Tracked entity instance: " + entityInstance.getUid()
-                    + " has multiple active enrollments in program: " + program.getUid() ).setReference( event.getEvent() ).incrementIgnored();
-            }
-
-            programInstance = programInstances.get( 0 );
-
-            if ( !programStage.getRepeatable() )
+            
+            if ( programInstance == null )
             {
-                programStageInstance = programStageInstanceService.getProgramStageInstance( programInstance,
-                    programStage );
-
-                if ( programStageInstance != null && !programStageInstance.getUid().equals( event.getEvent() ) )
+                List<ProgramInstance> programInstances = new ArrayList<>( programInstanceService.getProgramInstances( entityInstance, program, ProgramStatus.ACTIVE ) );
+                
+                if ( programInstances.isEmpty() )
                 {
-                    return new ImportSummary( ImportStatus.ERROR, "Program stage is not repeatable and an event already exists" )
-                        .setReference( event.getEvent() ).incrementIgnored();
+                    return new ImportSummary( ImportStatus.ERROR, "Tracked entity instance: " + entityInstance.getUid() + " is not enrolled in program: " + program.getUid() ).setReference( event.getEvent() ).incrementIgnored();
                 }
-            }
-            else
-            {
-                if ( !StringUtils.isEmpty( event.getEvent() ) )
+                else if ( programInstances.size() > 1 )
                 {
-                    programStageInstance = manager.getObject( ProgramStageInstance.class,
-                        importOptions.getIdSchemes().getProgramStageInstanceIdScheme(), event.getEvent() );
-
-                    if ( programStageInstance == null )
-                    {
-                        if ( !CodeGenerator.isValidUid( event.getEvent() ) )
-                        {
-                            return new ImportSummary( ImportStatus.ERROR, "Event.event did not point to a valid event: " + event.getEvent() )
-                                .setReference( event.getEvent() ).incrementIgnored();
-                        }
-                    }
+                    return new ImportSummary( ImportStatus.ERROR, "Tracked entity instance: " + entityInstance.getUid() + " has multiple active enrollments in program: " + program.getUid() ).setReference( event.getEvent() ).incrementIgnored();
                 }
+                
+                programInstance = programInstances.get( 0 );
```

Comment:
I am not sure I completely understand this logic. So we first check whether UID provided in the payload is present in DB (line 461). And if it is not, then we fetch all enrollments for given TEI and if it happens that there is exactly 1 (even though it can have completely different UID than the one provided in the payload) we will use it.

If I understand it correctly: Isn't it a bit weird? The user is trying to create a new event for some enrollment, but we will create it under completely different enrollment. Shouldn't we return an error?

Reply:
again this is a code which we have from before...

yes, first we check if the passed enrollment is present. If not, then we have two options - either create a new enrollment or fetch an existing active enrollment. We are not creating a new enrollment if the program is with registration. We only create for programs of without_registration.

What is happening there is fetch active enrollments for the given program and TEI - if there are more than 1 then we reject. if there is one we pick that and continue.